mips/intr.c

```diff
@@ -263,7 +265,31 @@ void kstack_overflow_handler(exc_frame_t *frame) {
     panic();
 }
 
-/* General exception handler is called with interrupts disabled. */
+/* Let's consider possible contexts that caused exception/interrupt:
+ *
+ * 1. We came from user-space:
+ *    interrupts and preemption must have been enabled
+ * 2. We came from kernel-space:
+ *    a. interrupts and preemption are enabled:
+ *       kernel was running in regular thread context
+ *    b. preemption is disabled, interrupts are enabled:
+ *       kernel was running in thread context and acquired sleep lock (mutex)
```

Comment:
Case 2b. Being an owner of a mutex is not related to having preemption disabled IMO. The description should be something like: `kernel was running in thread context and disabled preemption`.

Reply:
Right, but that in most cases this is a consequence of acquiring a sleep lock. I'll change it to inform reader that context is somewhat wider that implied by the comment.

Reply:
> Right, but that in most cases this is a consequence of acquiring a sleep lock.

How is that? We disable preemption in `sys/mutex.c` only for the time of execution of `turnstile` functions.
```
WITH_MTX_LOCK(&some_mtx) {
 // preemption is enabled here
}
```

Reply:
Ok. Let's consider following code from `mtx_lock`.

```c
WITH_NO_PREEMPTION {
    while (m->m_owner != NULL)
        turnstile_wait(m, (thread_t *)m->m_owner, waitpt); 
    m->m_owner = thread_self();
    m->m_lockpt = waitpt;
}
```

It definitely may get interrupted or an exception may happen somewhere inside. If we during exception / interrupt handling we allow to switch out to another code... boom – we broke the promise. Obviously switch out will not happen as a result of calling `on_exc_leave` – that's ruled out – but IMO still a programmer may call implicitely `ctx_switch`.

Reply:
I totally agree on the cases and the actions that should be taken in each of them. I only wanted to say that this comment:
```
 *    b. preemption is disabled, interrupts are enabled:
 *       kernel was running in thread context and acquired sleep lock (mutex)
```
is misleading as:
* preemption could've been disabled anywhere in the kernel, not only in `sys/mutex.c` code,
* IMO the comment implies that "thread has acquired a mutex" means "thread has disabled preemption" -- it should be stated that the preemption is disabled only during acquir**ing** and releas**ing** of the mutex.